File: src-tauri/src/services/claim_receipt/contracts.rs
//! Route-neutral DTO for surface-agnostic claim receipt rendering.
//!
//! Receipt is a render-side projection of existing claim/proposal state — it carries
//! no new substrate writes. Surfaces consume it through `render::render_receipt_for`.

use abilities_runtime::abilities::feedback::FeedbackAction;
use abilities_runtime::abilities::provenance::subject::SubjectRef;
use abilities_runtime::abilities::trust::types::TrustBand;
use abilities_runtime::sensitivity::{ClaimVerificationState, RenderableClaimText};
use abilities_runtime::types::{ClaimState, SurfacingState};
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ReceiptTarget {
    #[serde(rename_all = "camelCase")]
    Claim {
        claim_id: String,
        subject: SubjectRef,
        field_path: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    Proposal {
        proposal_id: String,
        subject: SubjectRef,
        field_path: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    WorkItem {
        action_id: String,
        backing_claim_id: Option<String>,
        subject: Option<SubjectRef>,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum SurfaceContext {
    ActionsWork,
    EntityDetail,
    DailyBriefing,
    MeetingDetail,
    Mcp,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Freshness {
    Current,
    Aging,
    Stale,
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum RedactionLevel {
    None,
    Partial,
    Full,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ReceiptTrust {
    pub band: TrustBand,
    pub source_asof: Option<DateTime<Utc>>,
    pub freshness: Freshness,
    pub caveat: Option<String>,
    pub rationale: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ReceiptLifecycle {
    pub claim_state: ClaimState,
    pub surfacing_state: SurfacingState,
    pub verification_state: ClaimVerificationState,
    pub updated_at: Option<DateTime<Utc>>,
}

/// Generic label; never the raw source id. Raw source identity is gated by render policy.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ProvenanceSource {
    pub label: String,
    pub source_type: Option<String>,
    pub as_of: Option<DateTime<Utc>>,
    pub href: Option<String>,
    pub redacted: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ReceiptProvenance {
    pub sources: Vec<ProvenanceSource>,
    pub field_path: Option<String>,
    pub evidence_summary: Option<String>,
    pub redaction: RedactionLevel,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ReceiptAction {
    pub action: FeedbackAction,
    pub label: String,
    pub disabled_reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ClaimReceipt {
    pub target: ReceiptTarget,
    pub surface_context: SurfaceContext,
    pub rendered_text: Option<RenderableClaimText>,
    pub trust: ReceiptTrust,
    pub lifecycle: ReceiptLifecycle,
    pub provenance: ReceiptProvenance,
    pub actions: Vec<ReceiptAction>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn receipt_target_serde_tagged_camel_case() {
        let target = ReceiptTarget::Claim {
            claim_id: "c1".into(),
            subject: SubjectRef::Global,
            field_path: Some("status".into()),
        };
        let json = serde_json::to_value(&target).unwrap();
        assert_eq!(json["kind"], "claim");
        assert!(json.get("claimId").is_some());
        assert!(json.get("fieldPath").is_some());
        let round: ReceiptTarget = serde_json::from_value(json).unwrap();
        assert_eq!(round, target);
    }

    #[test]
    fn surface_context_serde_snake_case() {
        let json = serde_json::to_string(&SurfaceContext::ActionsWork).unwrap();
        assert_eq!(json, "\"actions_work\"");
        let json = serde_json::to_string(&SurfaceContext::DailyBriefing).unwrap();
        assert_eq!(json, "\"daily_briefing\"");
    }

    #[test]
    fn provenance_redacted_flag_present() {
        let src = ProvenanceSource {
            label: "internal note".into(),
            source_type: None,
            as_of: None,
            href: None,
            redacted: true,
        };
        let json = serde_json::to_value(&src).unwrap();
        assert_eq!(json["redacted"], true);
    }

    #[test]
    fn receipt_action_carries_feedback_action() {
        let act = ReceiptAction {
            action: FeedbackAction::ConfirmCurrent,
            label: "Looks right".into(),
            disabled_reason: None,
        };
        let json = serde_json::to_value(&act).unwrap();
        assert_eq!(json["action"], "confirm_current");
        assert_eq!(json["label"], "Looks right");
    }
}

File: src/services/claim-receipt/contracts.ts
export type SubjectRef =
  | { account: string }
  | { project: string }
  | { person: string }
  | { meeting: string }
  | { user: string }
  | "global"
  | { multi: SubjectRef[] }
  | "unknown";

export type ReceiptTarget =
  | {
      kind: "claim";
      claimId: string;
      subject: SubjectRef;
      fieldPath?: string | null;
    }
  | {
      kind: "proposal";
      proposalId: string;
      subject: SubjectRef;
      fieldPath?: string | null;
    }
  | {
      kind: "workItem";
      actionId: string;
      backingClaimId?: string | null;
      subject?: SubjectRef | null;
    };

export type SurfaceContext =
  | "actions_work"
  | "entity_detail"
  | "daily_briefing"
  | "meeting_detail"
  | "mcp";

export type Freshness = "current" | "aging" | "stale" | "unknown";

export type RedactionLevel = "none" | "partial" | "full";

export type TrustBand =
  | "likely_current"
  | "use_with_caution"
  | "needs_verification"
  | "unscored";

export type ClaimState = "active" | "dormant" | "tombstoned" | "withdrawn";

export type SurfacingState = "active" | "dormant";

export type ClaimVerificationState = "active" | "contested" | "needs_user_decision";

export type FeedbackAction =
  | "confirm_current"
  | "mark_outdated"
  | "mark_false"
  | "wrong_subject"
  | "wrong_source"
  | "cannot_verify"
  | "needs_nuance"
  | "surface_inappropriate"
  | "not_relevant_here"
  | "merge_intent";

export type ClaimSensitivity = "public" | "internal" | "confidential" | "user_only";

export type RenderSurface =
  | "tauri_entity_detail"
  | "tauri_briefing_prep"
  | "tauri_meeting_detail"
  | "tauri_email_summary"
  | "action"
  | "tauri_provenance"
  | "tauri_report"
  | "tauri_chat"
  | "mcp_tool"
  | "mcp_tool_detail"
  | "p2_publication"
  | "log_structured"
  | "push_notification";

export type RenderPolicyKind = "render" | "redacted" | "drop";

export type RedactionAffordance =
  | {
      kind: "confidential_click_to_reveal";
      claim_id: string;
      label: string;
      audit_required: boolean;
    }
  | {
      kind: "confidential_hidden";
      label: string;
    }
  | {
      kind: "user_only_hidden";
      label: string;
    };

export interface RenderPolicy {
  kind: RenderPolicyKind;
  sensitivity: ClaimSensitivity;
  surface: RenderSurface;
  claimId?: string | null;
  affordance?: RedactionAffordance | null;
}

export interface RenderableClaimText {
  text: string;
  policy: RenderPolicy;
}

export interface ReceiptTrust {
  band: TrustBand;
  sourceAsof?: string | null;
  freshness: Freshness;
  caveat?: string | null;
  rationale?: string | null;
}

export interface ReceiptLifecycle {
  claimState: ClaimState;
  surfacingState: SurfacingState;
  verificationState: ClaimVerificationState;
  updatedAt?: string | null;
}

export interface ProvenanceSource {
  label: string;
  sourceType?: string | null;
  asOf?: string | null;
  href?: string | null;
  redacted: boolean;
}

export interface ReceiptProvenance {
  sources: ProvenanceSource[];
  fieldPath?: string | null;
  evidenceSummary?: string | null;
  redaction: RedactionLevel;
}

export interface ReceiptAction {
  action: FeedbackAction;
  label: string;
  disabledReason?: string | null;
}

export interface ClaimReceipt {
  target: ReceiptTarget;
  surfaceContext: SurfaceContext;
  renderedText?: RenderableClaimText | null;
  trust: ReceiptTrust;
  lifecycle: ReceiptLifecycle;
  provenance: ReceiptProvenance;
  actions: ReceiptAction[];
}
